crates/router-controller/src/controller.rs

```diff
@@ -125,10 +129,39 @@ impl Controller {
 
         match candidates.len() {
             0 => None,
-            1 => Some(&candidates[0].url),
+            1 => {
+                let worker = &candidates[0];
+                self.workers_rate.inc(&worker.url);
+                Some(&worker.url)
+            },
             len => {
-                let i: usize = rand::random();
-                Some(&candidates[i % len].url)
+                let key = (dataset.clone(), first_block);
+
+                let next_worker = if let Some(url) = self.request_tracker.get(&key) {
```

Comment:
Is it used for caching recently sent requests? Should it be renamed to `request_cache` then?